### DISCUSSION TITLE
Capture the output of a shell command?

### DISCUSSION THREAD

**Discussion by uvtc** [janet-lang/janet#654] [Q&A]
How could I capture the output of a shell command? As in something like `(let [d (os/shell "date")] ...)` (where `d` would then be that string that the `date` shell command would normally print out)?

I also tried `(os/execute ["date"] :p)`, but again it just prints its output to stdout.

**Comment by pepe** [comment]
I think the best way is with `os/spawn` as shown in the first example https://janetdocs.com/os%2fspawn

**Reply by uvtc** [reply]
Thanks. Ok, I try that and get:

~~~
repl:1:> (def x (os/spawn ["date"] :p {:out :pipe}))
<core/process 0x55EF40AB5080>
repl:2:> x
<core/process 0x55EF40AB5080>
~~~

I don't want a process though. I want it to run and collect its output. I want (for this example) `x` to be some date string from the `date` command.

**Reply by uvtc** [reply]
Ooof. Ok, I see the rest now.

~~~
repl:1:> (def x (os/spawn ["date"] :p {:out :pipe}))
<core/process 0x55BA26D57080>

repl:2:> (def s (:read (x :out) :all))
@"Tue 09 Mar 2021 03:04:41 PM EST\n"

repl:3:> s
@"Tue 09 Mar 2021 03:04:41 PM EST\n"

repl:4:> (:wait x)
0
~~~

**Reply by uvtc** [reply]
@pepe , so, turning that into a function, and with some help from @andrewchambers , I've got:

~~~clojure
(defn shell-out
  [cmd]
  (let [x (os/spawn cmd :p {:out :pipe})
        s (:read (x :out) :all)]
    (:wait x)
    s))

(pp (shell-out ["date"]))
(pp "---------")
(pp (shell-out ["find" "." "-name" "*.txt"]))
~~~

which seems to work for those two examples.

Not sure I'm naming that right though.

**Reply by pepe** [reply]
I guess the name is as good as any. Glad it helped.

**Reply by uvtc** [reply]
I mention the name because, I remember from Perl that you could use backticks to execute code and get its output (as in ``` my $d = `date`; ```), and I think that was the same thing as calling the `system` function. (Doesn't look like Janet has a function specifically named "system" or "os/system".)

Thanks, @pepe .
